pkg/morph/client/nns.go

```diff
@@ -107,76 +102,29 @@ func nnsResolveItem(c *rpcclient.WSClient, nnsHash util.Uint160, domain string)
 		return nil, ErrNNSRecordNotFound
 	}
 
-	result, err := c.InvokeFunction(nnsHash, "resolve", []smartcontract.Parameter{
-		{
-			Type:  smartcontract.StringType,
-			Value: domain,
-		},
-		{
-			Type:  smartcontract.IntegerType,
-			Value: big.NewInt(int64(nns.TXT)),
-		},
-	}, nil)
+	result, err := nnsReader.Resolve(domain, nns.TXT)
 	if err != nil {
 		return nil, err
 	}
-	if result.State != vmstate.Halt.String() {
-		return nil, fmt.Errorf("invocation failed: %s", result.FaultException)
-	}
-	if len(result.Stack) == 0 {
+	if len(result) == 0 {
 		return nil, errEmptyResultStack
 	}
-	return result.Stack[0], nil
-}
-
-func nnsResolve(c *rpcclient.WSClient, nnsHash util.Uint160, domain string) (util.Uint160, error) {
-	res, err := nnsResolveItem(c, nnsHash, domain)
-	if err != nil {
-		return util.Uint160{}, err
-	}
-
-	// Parse the result of resolving NNS record.
-	// It works with multiple formats (corresponding to multiple NNS versions).
-	// If array of hashes is provided, it returns only the first one.
-	if arr, ok := res.Value().([]stackitem.Item); ok {
-		if len(arr) == 0 {
-			return util.Uint160{}, errors.New("NNS record is missing")
-		}
-		res = arr[0]
-	}
-	bs, err := res.TryBytes()
-	if err != nil {
-		return util.Uint160{}, fmt.Errorf("malformed response: %w", err)
-	}
-
-	// We support several formats for hash encoding, this logic should be maintained in sync
-	// with parseNNSResolveResult from cmd/neofs-adm/internal/modules/morph/initialize_nns.go
-	h, err := util.Uint160DecodeStringLE(string(bs))
-	if err == nil {
-		return h, nil
-	}
-
-	h, err = address.StringToUint160(string(bs))
-	if err == nil {
-		return h, nil
-	}
-
-	return util.Uint160{}, errors.New("no valid hashes are found")
+	return result, nil
 }
 
 // exists checks domain presence in the NNS contract with given address by
 // calling `ownerOf` method. Returns true if call succeeded only.
-func exists(c *rpcclient.WSClient, nnsHash util.Uint160, domain string) (bool, error) {
-	result, err := c.InvokeFunction(nnsHash, "ownerOf", []smartcontract.Parameter{
-		{
-			Type:  smartcontract.ByteArrayType,
-			Value: []byte(domain),
-		},
-	}, nil)
+func exists(nnsReader *nns.ContractReader, domain string) (bool, error) {
+	_, err := nnsReader.OwnerOf([]byte(domain))
+
 	if err != nil {
+		if strings.Contains(err.Error(), "invocation failed") {
```

Comment:
is it the best we can suggest here?

Reply:
Until https://github.com/nspcc-dev/neo-go/issues/3130 yes.